File: 347-top-k-frequent-elements/347. Top K Frequent Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq_map = Counter(nums)
        bucket = [[] for _ in range(len(nums) + 1)]
        for num, freq in freq_map.items():
            bucket[freq].append(num)
        flat_list = [item for arr in bucket for item in arr]
        print(bucket)
        print(flat_list)

        return flat_list[::-1][:k]

File: 347-top-k-frequent-elements/347. Top K Frequent Elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freq;
        for (int n : nums) {
            freq[n]++;
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
        vector<int> res;
        for (auto [n, f] : freq) {
            if (minHeap.size() < k) {
                minHeap.push({f, n});
            } else {
                if (f > minHeap.top().first) {
                    minHeap.pop();
                    minHeap.push({f, n});
                }
            }
        }
        while (!minHeap.empty()) {
            res.push_back(minHeap.top().second);
            minHeap.pop();
        }
        return res; 
    }
};